packages/fork-choice/src/forkChoice/forkChoice.ts

```diff
@@ -162,25 +176,42 @@ export class ForkChoice implements IForkChoice {
 
     let timer;
     this.metrics?.forkChoiceRequests.inc();
-
     try {
+      let deltas: number[];
+
+      // Check if scores need to be calculated/updated
       if (!this.synced) {
         timer = this.metrics?.forkChoiceFindHead.startTimer();
-        const deltas = computeDeltas(
-          this.protoArray.indices,
-          this.votes,
-          this.justifiedBalances,
-          this.justifiedBalances
-        );
+        deltas = computeDeltas(this.protoArray.indices, this.votes, this.justifiedBalances, this.justifiedBalances);
+        /**
+         * The structure in line with deltas to propogate boost up the branch
+         * starting from the proposerIndex
+         */
+        const boosts: number[] = [];
+        if (this.proposerBoostEnabled && this.proposerBoost) {
+          const proposerIndex = this.protoArray.indices.get(this.proposerBoost.root);
+          if (proposerIndex == undefined) throw Error("InvalidProposerIndex");
+          const proposerBoostScore =
+            this.proposerBoost.score ??
+            computeProposerBoostScoreFromBalances(this.justifiedBalances, {
```

Comment:
this function should be called once per justified balances at most, we can even precompute it in `getEffectiveBalances()` to avoid an extra `for` loop over justified balances, and send both `proposerBoostScore` and `justifiedBalances` to forkchoice over `onBlock()`

Reply:
yes, good idea :+1:

Reply:
actually it could require a bit more bookeeping as there is one place where justified is updated with the bestJustifiedBalances, so will also have to track bestProposerBoostScore

Reply:
@tuyennhv now score is calculated atmost once per justified balances